inc/zoo/swar/SWAR.h

```diff
@@ -83,99 +88,153 @@ namespace detail {
+template<int NBits, typename T = uint64_t> constexpr auto horizontalEquality(SWAR<NBits, T> left, SWAR<NBits, T> right) {
+  return left.m_v == right.m_v;
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto isolate(T pattern) {
+  return pattern & ((T(1)<<NBits)-1);
+}
+
+template<typename T = uint64_t>
+constexpr auto clearLSB(T v) {
+  return v & (v - 1);
+}
+
+template<typename T = uint64_t>
+constexpr auto isolateLSB(T v) {
+  return v & ~clearLSB(v);
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto maskLowBits() {
+  return (T(1)<<(NBits-1)) | ((T(1)<<(NBits-1))-1);
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto clearLSBits(T v) {
+  return v &(~(maskLowBits<NBits>() << metaLogFloor(isolateLSB<T>(v))));
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto isolateLSBits(T v) {
+  return v &(maskLowBits<NBits>() << metaLogFloor(isolateLSB<T>(v)));
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto broadcast(SWAR<NBits, T> v) {
+  constexpr T Ones = makeBitmask<NBits, T>(SWAR<NBits, T>{1}.value());
+  return SWAR<NBits, T>(T(v) * Ones);
+}
+
+template<int NBits, typename T>
+struct BooleanSWAR: SWAR<NBits, T> {
+    // Booleanness is stored in MSB of a given swar.
+    static constexpr T MSBs = broadcast<NBits, T>(SWAR<NBits, T>(T(1) << (NBits -1))).value();
+    constexpr explicit BooleanSWAR(T v): SWAR<NBits, T>(v) {}
 
-    constexpr BooleanSWAR clear(int bit) {
-        constexpr auto Bit = T(1) << (Size - 1);
-        return this->m_v ^ (Bit << (Size * bit)); }
-    constexpr int best() { return this->top(); }
+    constexpr BooleanSWAR clear(int bit) const noexcept {
+        constexpr auto Bit = T(1) << (NBits - 1);
+        return this->m_v ^ (Bit << (NBits * bit)); }
+    constexpr int best() const noexcept { return this->top(); }
+
+    constexpr operator bool() const noexcept { return this->m_v; }
+
+    constexpr BooleanSWAR operator not() const noexcept {
+        return MSBs ^ *this;
+    }
+ private:
+    constexpr BooleanSWAR(SWAR<NBits, T> initializer) noexcept: SWAR<NBits, T>(initializer) {}
+
+    template<int NB, typename TT> friend
+    constexpr BooleanSWAR<NB, TT> greaterEqual_MSB_off(SWAR<NB, TT> left, SWAR<NB, TT> right) noexcept;
 
-    constexpr operator bool() const { return this->m_v; }
 };
 
-template<int N, int Size, typename T>
-constexpr BooleanSWAR<Size, T> greaterEqualSWAR(SWAR<Size, T> v) {
-    static_assert(1 < Size, "Degenerated SWAR");
-    static_assert(metaLogCeiling(N) < Size, "N is too big for this technique");
-    constexpr auto msbPos  = Size - 1;
+template<int NBits, typename T>
+constexpr BooleanSWAR<NBits, T> greaterEqual_MSB_off(SWAR<NBits, T> left, SWAR<NBits, T> right) noexcept {
+    constexpr auto MSOnes = BooleanSWAR<NBits, T>::MSBs;
+    return SWAR<NBits, T>{
+        (((left.value() | MSOnes) - right.value()) & MSOnes)
+    };
```

Comment:
the `values()` projection is not needed, since you should be able to rely on the overloaded operators for `SWAR`

Reply:
I must have removed those operators, or am catching an implicit conversion somewhere I missed. Let me check ...

Reply:
operator- isn't on swar, will add next PR (well, adjust to xlist and add operator-), leaving this for now.